python/cloudtik/providers/_private/huaweicloud/utils.py

```diff
@@ -0,0 +1,96 @@
+import logging
+from functools import lru_cache
+from typing import Any, Dict
+
+from huaweicloudsdkcore.auth.credentials import BasicCredentials
+from huaweicloudsdkcore.http.http_config import HttpConfig
+from huaweicloudsdkecs.v2 import EcsClient
+from huaweicloudsdkecs.v2.region.ecs_region import EcsRegion
+from huaweicloudsdkeip.v2 import EipClient
+from huaweicloudsdkeip.v2.region.eip_region import EipRegion
+from huaweicloudsdknat.v2 import NatClient
+from huaweicloudsdknat.v2.region.nat_region import NatRegion
+from huaweicloudsdkvpc.v2 import VpcClient
+from huaweicloudsdkvpc.v2.region.vpc_region import VpcRegion
+
+from cloudtik.core._private.constants import env_bool
+
+logger = logging.getLogger(__name__)
+
+
+@lru_cache()
+def _client_cache(ak: str, sk: str, region: str) -> Dict[str, Any]:
+    client_map = {}
+    credentials = BasicCredentials(ak, sk)
+
+    # Get proxy setting, if $HWC_IGNORE_SSL_VERIFICATION is true explicitly,
+    # ignore checking, in other case enable SSL verifying.
+    http_config = HttpConfig.get_default_config()
+    http_config.ignore_ssl_verification = env_bool(
+        'HWC_IGNORE_SSL_VERIFICATION', False)
+
+    ecs_client = EcsClient.new_builder() \
+        .with_http_config(http_config) \
+        .with_credentials(credentials) \
+        .with_region(EcsRegion.value_of(region)) \
+        .build()
+    client_map['ecs'] = ecs_client
+
+    vpc_client = VpcClient.new_builder() \
+        .with_http_config(http_config) \
+        .with_credentials(credentials) \
+        .with_region(VpcRegion.value_of(region)) \
+        .build()
+    client_map['vpc'] = vpc_client
+
+    nat_client = NatClient.new_builder() \
+        .with_http_config(http_config) \
+        .with_credentials(credentials) \
+        .with_region(NatRegion.value_of(region)) \
+        .build()
+    client_map['nat'] = nat_client
+
+    eip_client = EipClient.new_builder() \
+        .with_http_config(http_config) \
+        .with_credentials(credentials) \
+        .with_region(EipRegion.value_of(region)) \
+        .build()
+    client_map['eip'] = eip_client
+
+    return client_map
+
+
+def make_ecs_client(config: Dict[str, Any]) -> Any:
+    config_provider = config['provider']
+    credentials = config_provider['huaweicloud_credentials']
+    ak = credentials['huaweicloud_access_key']
+    sk = credentials['huaweicloud_secret_key']
+    region = config_provider['region']
+    return _client_cache(ak, sk, region)['ecs']
```

Comment:
For making a client for workspace (which is run at user's working machine), we suggest to consider multiple options user may prefer for the specific cloud.
1. the access key and secret key specified explicitly in the huaweicloud_credentials.
2. If there is no huaweicloud_credentials, utilize something setup in the environment including environment variable or instance credentials (support to use in node provider)

The handling logic is when there is explicit huaweicloud_credentials specified, the code we try use the credentials existing in the environment. 

This way will be robust for using both at working node (with user may setup the environment using cloud CLI) and using (the ecs client) at the Node provider  which will be run both on working node (workspace and cluster booting) and on head node (scaling workers, which will use instance credentials instead of passing the secret keys in config file). 'huaweicloud_credentials' will be removed from config when passing the config to head to avoid leak any secret keys in config.

We can follow the same credential pattern for making other clients.

Reply:
aha, this use case make sense, I did't consider it yet, will add support in next commit